In [1]:
import torch
from torch import nn
import torch.nn.functional as F

## Homework 1 - PyTorch basics
*Lorenzo Basile*

#### 1, 2

In [2]:
class MLP(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.layer1 = nn.Linear(in_features=5, out_features=11, bias=False)
        self.layer2 = nn.Linear(in_features=11, out_features=16, bias=False)
        self.layer3 = nn.Linear(in_features=16, out_features=13, bias=False)
        self.layer4 = nn.Linear(in_features=13, out_features=8, bias=False)
        self.layer5 = nn.Linear(in_features=8, out_features=4, bias=False)
    def forward(self, x):
        out = F.relu(self.layer1(x))
        out = F.relu(self.layer2(out))
        out = F.relu(self.layer3(out))
        out = F.relu(self.layer4(out))
        return F.softmax(self.layer5(out))

In [3]:
net=MLP()
print("Number of parameters: ", sum(p.numel() for p in net.parameters()))
print(net)

Number of parameters:  575
MLP(
  (layer1): Linear(in_features=5, out_features=11, bias=False)
  (layer2): Linear(in_features=11, out_features=16, bias=False)
  (layer3): Linear(in_features=16, out_features=13, bias=False)
  (layer4): Linear(in_features=13, out_features=8, bias=False)
  (layer5): Linear(in_features=8, out_features=4, bias=False)
)


#### 3 

In a MLP, if biases are not present, the number of parameters (weights) for layer $k$ ($k\neq 0$) is $N_{k-1}\cdot N_k$.

So, since the layers have sizes $N=[5, 11, 16, 13, 8, 4]$ the total parameters are:
$$
\#weights=5\cdot 11 + 11 \cdot 16 + 16 \cdot 13 + 13 \cdot 8 + 8 \cdot 4 = 575
$$

The introduction of biases in all layers would lead to an increase in the number of parameters by a quantity equal to the total number of neurons in the MLP (excluding input neurons).
So, the total number of parameters would become:
$$
\#params=\#weights+11+16+13+8+4=627
$$

#### 4

In [4]:
for p in net.parameters():
    print (torch.linalg.norm(p.detach(), 2), torch.linalg.norm(p.detach(), 1))

tensor(1.1866) tensor(2.9274)
tensor(1.2086) tensor(3.1139)
tensor(0.9356) tensor(1.9889)
tensor(0.8469) tensor(1.6046)
tensor(0.7770) tensor(1.0768)
